File: master_wallet/flutter/lib/services/web_socket_service.dart
/**
 * WebSocketService - Flutter Implementation
 * Real-time connection for Master Wallet
 */

import 'dart:async';
import 'dart:convert';
import 'package:web_socket_channel/web_socket_channel.dart';

enum ConnectionState {
  disconnected,
  connecting,
  connected,
  reconnecting,
  error,
}

class WebSocketService {
  static final WebSocketService _instance = WebSocketService._internal();
  factory WebSocketService() => _instance;
  WebSocketService._internal();
  
  WebSocketChannel? _channel;
  StreamSubscription? _subscription;
  Timer? _heartbeatTimer;
  Timer? _reconnectTimer;
  
  String? _walletId;
  String? _authToken;
  int _reconnectAttempts = 0;
  
  static const String WS_HOST = String.fromEnvironment(
    'MASTER_WALLET_WS_URL',
    defaultValue: 'ws://localhost:8450',
  );
  static const int MAX_RECONNECT_ATTEMPTS = 10;
  static const Duration RECONNECT_DELAY = Duration(seconds: 5);
  
  ConnectionState _connectionState = ConnectionState.disconnected;
  ConnectionState get connectionState => _connectionState;
  
  final _stateController = StreamController<ConnectionState>.broadcast();
  Stream<ConnectionState> get stateStream => _stateController.stream;
  
  final _messageController = StreamController<String>.broadcast();
  Stream<String> get messageStream => _messageController.stream;
  
  final _balanceController = StreamController<BalanceUpdate>.broadcast();
  Stream<BalanceUpdate> get balanceStream => _balanceController.stream;
  
  final _transactionController = StreamController<TransactionUpdate>.broadcast();
  Stream<TransactionUpdate> get transactionStream => _transactionController.stream;
  
  /// Connect to WebSocket server
  void connect({required String walletId, String? token}) {
    _walletId = walletId;
    _authToken = token;
    _connect();
  }
  
  void _connect() {
    _connectionState = ConnectionState.connecting;
    _stateController.add(_connectionState);
    
    try {
      // Canonical WebSocket: GET /ws?master_wallet_id=&token=
      final qs = <String>[];
      if (_walletId != null && _walletId!.isNotEmpty) {
        qs.add('master_wallet_id=${Uri.encodeQueryComponent(_walletId!)}');
      }
      if (_authToken != null && _authToken!.isNotEmpty) {
        qs.add('token=${Uri.encodeQueryComponent(_authToken!)}');
      }
      final url = qs.isEmpty ? '$WS_HOST/ws' : '$WS_HOST/ws?${qs.join('&')}';
      _channel = WebSocketChannel.connect(Uri.parse(url));
      
      _subscription = _channel!.stream.listen(
        _onMessage,
        onError: _onError,
        onDone: _onDone,
      );
      
      _connectionState = ConnectionState.connected;
      _stateController.add(_connectionState);
      _reconnectAttempts = 0;
      
      _authenticate();
      _startHeartbeat();
    } catch (e) {
      _connectionState = ConnectionState.error;
      _stateController.add(_connectionState);
      _scheduleReconnect();
    }
  }
  
  /// Disconnect from server
  void disconnect() {
    _stopHeartbeat();
    _cancelReconnect();
    _subscription?.cancel();
    _channel?.sink.close();
    _channel = null;
    
    _connectionState = ConnectionState.disconnected;
    _stateController.add(_connectionState);
  }
  
  /// Subscribe to balance updates
  void subscribeToBalance(int chainId) {
    _sendMessage('subscribe', 'balance', {'chainId': chainId});
  }
  
  /// Unsubscribe from balance updates
  void unsubscribeFromBalance(int chainId) {
    _sendMessage('unsubscribe', 'balance', {'chainId': chainId});
  }
  
  /// Subscribe to transaction updates
  void subscribeToTransactions(String address) {
    _sendMessage('subscribe', 'transactions', {'address': address});
  }
  
  /// Subscribe to ticker updates
  void subscribeToTicker(String pair) {
    _sendMessage('subscribe', 'ticker', {'pair': pair});
  }
  
  /// Subscribe to order book
  void subscribeToOrderBook(String pair) {
    _sendMessage('subscribe', 'orderbook', {'pair': pair});
  }
  
  void _authenticate() {
    _sendMessage('auth', 'auth', {
      'walletId': _walletId ?? '',
      'token': _authToken ?? '',
    });
  }
  
  void _sendMessage(String type, String channel, Map<String, dynamic> data) {
    final message = {
      'type': type,
      'channel': channel,
      'data': data,
      'timestamp': DateTime.now().millisecondsSinceEpoch,
    };
    
    _channel?.sink.add(jsonEncode(message));
  }
  
  void _onMessage(dynamic data) {
    try {
      final json = jsonDecode(data as String) as Map<String, dynamic>;
      _messageController.add(data as String);

      // The canonical backend emits flat messages keyed by `type`
      // (e.g. "balance", "ticker", "tx_confirmation"), not channel/data envelopes.
      final type = json['type'] as String?;
      switch (type) {
        case 'balance':
          _handleBalanceUpdate(json);
          break;
        case 'tx_confirmation':
        case 'transaction':
          _handleTransactionUpdate(json);
          break;
      }
    } catch (e) {
      // Ignore malformed frames; the connection itself is retried elsewhere.
    }
  }

  void _handleBalanceUpdate(Map<String, dynamic> data) {
    _balanceController.add(BalanceUpdate(
      chainId: _asInt(data['chain_id'] ?? data['chainId']) ?? 0,
      address: data['address'] as String? ?? '',
      balance: data['balance']?.toString() ?? '0',
      token: data['symbol'] as String? ?? data['token'] as String? ?? 'ETH',
      timestamp: _asInt(data['timestamp']) ??
          DateTime.now().millisecondsSinceEpoch,
    ));
  }

  void _handleTransactionUpdate(Map<String, dynamic> data) {
    _transactionController.add(TransactionUpdate(
      txHash: data['tx_hash'] as String? ?? data['txHash'] as String? ?? '',
      from: data['from'] as String? ?? '',
      to: data['to'] as String? ?? '',
      amount: data['amount']?.toString() ?? '0',
      status: data['status'] as String? ?? '',
      timestamp: _asInt(data['timestamp']) ??
          DateTime.now().millisecondsSinceEpoch,
    ));
  }

  static int? _asInt(dynamic v) {
    if (v == null) return null;
    if (v is int) return v;
    if (v is num) return v.toInt();
    return int.tryParse(v.toString());
  }
  
  void _onError(dynamic error) {
    _connectionState = ConnectionState.error;
    _stateController.add(_connectionState);
    _stopHeartbeat();
    _scheduleReconnect();
  }
  
  void _onDone() {
    _connectionState = ConnectionState.disconnected;
    _stateController.add(_connectionState);
    _stopHeartbeat();
    _scheduleReconnect();
  }
  
  void _scheduleReconnect() {
    if (_reconnectAttempts >= MAX_RECONNECT_ATTEMPTS) {
      _connectionState = ConnectionState.error;
      _stateController.add(_connectionState);
      return;
    }
    
    _reconnectAttempts++;
    _connectionState = ConnectionState.reconnecting;
    _stateController.add(_connectionState);
    
    _reconnectTimer = Timer(
      RECONNECT_DELAY * _reconnectAttempts,
      _connect,
    );
  }
  
  void _cancelReconnect() {
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
  }
  
  void _startHeartbeat() {
    _heartbeatTimer = Timer.periodic(const Duration(seconds: 15), (_) {
      _sendMessage('ping', 'heartbeat', {});
    });
  }
  
  void _stopHeartbeat() {
    _heartbeatTimer?.cancel();
    _heartbeatTimer = null;
  }
  
  void dispose() {
    disconnect();
    _stateController.close();
    _messageController.close();
    _balanceController.close();
    _transactionController.close();
  }
}

// Data Classes

class BalanceUpdate {
  final int chainId;
  final String address;
  final String balance;
  final String token;
  final int timestamp;
  
  BalanceUpdate({
    required this.chainId,
    required this.address,
    required this.balance,
    required this.token,
    required this.timestamp,
  });
}

class TransactionUpdate {
  final String txHash;
  final String from;
  final String to;
  final String amount;
  final String status;
  final int timestamp;
  
  TransactionUpdate({
    required this.txHash,
    required this.from,
    required this.to,
    required this.amount,
    required this.status,
    required this.timestamp,
  });
}

File: master_wallet/web/src/services/webSocketService.ts
/**
 * WebSocketService - Web (React/TypeScript)
 * Real-time connection to the canonical backend WebSocket at ws://localhost:8450/ws
 * (derived from MASTER_WALLET_API_URL). Sends master_wallet_id + token query params
 * per the canonical contract; no loopback fakes.
 */

import { masterWalletAPI, getAuthToken } from '../api';

type ConnectionState = 'disconnected' | 'connecting' | 'connected' | 'reconnecting' | 'error';

interface BalanceUpdate {
  chainId: number;
  address: string;
  balance: string;
  token: string;
  timestamp: number;
}

interface TransactionUpdate {
  txHash: string;
  from: string;
  to: string;
  amount: string;
  status: string;
  timestamp: number;
}

class WebSocketService {
  private ws: WebSocket | null = null;
  private walletId: string | null = null;
  private authToken: string | null = null;
  private reconnectAttempts = 0;
  private heartbeatInterval: ReturnType<typeof setInterval> | null = null;
  private reconnectTimeout: ReturnType<typeof setTimeout> | null = null;

  private static readonly MAX_RECONNECT_ATTEMPTS = 10;
  private static readonly RECONNECT_DELAY = 5000;

  private _connectionState: ConnectionState = 'disconnected';
  get connectionState(): ConnectionState { return this._connectionState; }

  private stateListeners: Set<(state: ConnectionState) => void> = new Set();
  private messageListeners: Set<(message: string) => void> = new Set();
  private balanceListeners: Set<(update: BalanceUpdate) => void> = new Set();
  private transactionListeners: Set<(update: TransactionUpdate) => void> = new Set();

  private wsUrl(): string {
    return masterWalletAPI.wsUrl; // e.g. ws://localhost:8450/ws
  }

  /**
   * Connect to the canonical backend WebSocket. master_wallet_id + token are
   * passed as query params (the backend scopes broadcasts by master_wallet_id).
   */
  connect(walletId: string, token?: string): void {
    this.walletId = walletId;
    this.authToken = token ?? getAuthToken();
    this._connect();
  }

  private _connect(): void {
    this._connectionState = 'connecting';
    this._notifyStateChange();

    let url = this.wsUrl();
    const params = new URLSearchParams();
    if (this.walletId) params.set('master_wallet_id', this.walletId);
    if (this.authToken) params.set('token', this.authToken);
    const qs = params.toString();
    if (qs) url += `?${qs}`;

    try {
      this.ws = new WebSocket(url);

      this.ws.onopen = () => {
        this._connectionState = 'connected';
        this._notifyStateChange();
        this.reconnectAttempts = 0;
        this._startHeartbeat();
      };

      this.ws.onmessage = (event) => {
        this._handleMessage(event.data);
      };

      this.ws.onerror = (error) => {
        console.error('WebSocket error:', error);
        this._connectionState = 'error';
        this._notifyStateChange();
        this._stopHeartbeat();
        this._scheduleReconnect();
      };

      this.ws.onclose = () => {
        this._connectionState = 'disconnected';
        this._notifyStateChange();
        this._stopHeartbeat();
        this._scheduleReconnect();
      };
    } catch (error) {
      console.error('WebSocket connect failed:', error);
      this._connectionState = 'error';
      this._notifyStateChange();
      this._scheduleReconnect();
    }
  }

  /**
   * Disconnect from server
   */
  disconnect(): void {
    this._stopHeartbeat();
    this._cancelReconnect();

    if (this.ws) {
      this.ws.close(1000, 'Client disconnected');
      this.ws = null;
    }

    this._connectionState = 'disconnected';
    this._notifyStateChange();
  }

  /**
   * Subscribe to balance updates
   */
  subscribeToBalance(chainId: number): void {
    this._sendMessage('subscribe', 'balance', { chainId });
  }

  /**
   * Unsubscribe from balance updates
   */
  unsubscribeFromBalance(chainId: number): void {
    this._sendMessage('unsubscribe', 'balance', { chainId });
  }

  /**
   * Subscribe to transaction updates
   */
  subscribeToTransactions(address: string): void {
    this._sendMessage('subscribe', 'transactions', { address });
  }

  /**
   * Subscribe to ticker updates
   */
  subscribeToTicker(pair: string): void {
    this._sendMessage('subscribe', 'ticker', { pair });
  }

  /**
   * Subscribe to order book
   */
  subscribeToOrderBook(pair: string): void {
    this._sendMessage('subscribe', 'orderbook', { pair });
  }

  /**
   * Add state listener
   */
  onStateChange(listener: (state: ConnectionState) => void): void {
    this.stateListeners.add(listener);
  }

  /**
   * Remove state listener
   */
  offStateChange(listener: (state: ConnectionState) => void): void {
    this.stateListeners.delete(listener);
  }

  /**
   * Add message listener
   */
  onMessage(listener: (message: string) => void): void {
    this.messageListeners.add(listener);
  }

  /**
   * Add balance listener
   */
  onBalanceUpdate(listener: (update: BalanceUpdate) => void): void {
    this.balanceListeners.add(listener);
  }

  /**
   * Add transaction listener
   */
  onTransactionUpdate(listener: (update: TransactionUpdate) => void): void {
    this.transactionListeners.add(listener);
  }

  private _sendMessage(type: string, channel: string, data: Record<string, unknown>): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      return;
    }

    const message = {
      type,
      channel,
      data,
      timestamp: Date.now(),
    };

    this.ws.send(JSON.stringify(message));
  }

  private _handleMessage(data: string): void {
    try {
      const json = JSON.parse(data) as Record<string, unknown>;
      // The backend broadcasts messages with a `type` field (e.g. "balance",
      // "tx_confirmation", "ticker"). Route by `type`, falling back to
      // `channel` for client-originated subscriptions.
      const channel = (json.channel as string) ?? (json.type as string) ?? '';
      const payload = (json.data as Record<string, unknown>) ?? json;

      this.messageListeners.forEach((listener) => listener(data));

      switch (channel) {
        case 'balance':
          this._handleBalanceUpdate(payload);
          break;
        case 'tx_confirmation':
        case 'transactions':
          this._handleTransactionUpdate(payload);
          break;
        default:
          break;
      }
    } catch (error) {
      console.error('Error parsing message:', error);
    }
  }

  private _handleBalanceUpdate(data: Record<string, unknown>): void {
    const update: BalanceUpdate = {
      chainId: data.chain_id as number,
      address: data.address as string,
      balance: data.balance as string,
      token: (data.symbol as string) ?? (data.token as string) ?? '',
      timestamp: data.timestamp ? Number(data.timestamp) : Date.now(),
    };

    this.balanceListeners.forEach((listener) => listener(update));
  }

  private _handleTransactionUpdate(data: Record<string, unknown>): void {
    const update: TransactionUpdate = {
      txHash: (data.tx_hash as string) ?? (data.transaction_hash as string) ?? '',
      from: (data.from as string) ?? (data.from_address as string) ?? '',
      to: (data.to as string) ?? (data.to_address as string) ?? '',
      amount: data.amount as string,
      status: data.status as string,
      timestamp: data.timestamp ? Number(data.timestamp) : Date.now(),
    };

    this.transactionListeners.forEach((listener) => listener(update));
  }

  private _notifyStateChange(): void {
    this.stateListeners.forEach((listener) => listener(this._connectionState));
  }

  private _scheduleReconnect(): void {
    if (this.reconnectAttempts >= WebSocketService.MAX_RECONNECT_ATTEMPTS) {
      this._connectionState = 'error';
      this._notifyStateChange();
      return;
    }

    this.reconnectAttempts++;
    this._connectionState = 'reconnecting';
    this._notifyStateChange();

    this.reconnectTimeout = setTimeout(
      () => this._connect(),
      WebSocketService.RECONNECT_DELAY * this.reconnectAttempts
    );
  }

  private _cancelReconnect(): void {
    if (this.reconnectTimeout) {
      clearTimeout(this.reconnectTimeout);
      this.reconnectTimeout = null;
    }
  }

  private _startHeartbeat(): void {
    this.heartbeatInterval = setInterval(() => {
      this._sendMessage('ping', 'heartbeat', {});
    }, 15000);
  }

  private _stopHeartbeat(): void {
    if (this.heartbeatInterval) {
      clearInterval(this.heartbeatInterval);
      this.heartbeatInterval = null;
    }
  }
}

export const webSocketService = new WebSocketService();
export default webSocketService;
